cdap-data-fabric/src/main/java/co/cask/cdap/store/OwnerStore.java

```diff
@@ -0,0 +1,204 @@
+ * Refer to <a href=https://tools.ietf.org/html/rfc4120#section-7.5.8>Name Types</a> documentation for details on
+ * Kerberos Name Types.
+ * </p>
+ */
+public class OwnerStore {
+  private static final String OWNER_PREFIX = "o";
+  // currently, we only leverage one column of the table. However, not using KeyValueTable, so that being able to use
+  // additional columns in the future is simple
+  private static final byte[] COL = Bytes.toBytes("c");
+  private static final DatasetId DATASET_ID = NamespaceId.SYSTEM.dataset("owner.meta");
+  private static final DatasetProperties DATASET_PROPERTIES =
+    DatasetProperties.builder().add(Table.PROPERTY_CONFLICT_LEVEL, ConflictDetection.COLUMN.name()).build();
+
+  private final DatasetFramework datasetFramework;
+  private final Transactional transactional;
+
+  @Inject
+  OwnerStore(DatasetFramework datasetFramework, TransactionSystemClient txClient) {
+    this.datasetFramework = datasetFramework;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(new SystemDatasetInstantiator(datasetFramework),
+                                                                   txClient, DATASET_ID.getParent(),
+                                                                   Collections.<String, String>emptyMap(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+  }
+
+  /**
+   * Adds datasets and types to the given {@link DatasetFramework} used by {@link OwnerStore}
+   *
+   * @param framework framework to add types and datasets to
+   */
+  public static void setupDatasets(DatasetFramework framework) throws IOException, DatasetManagementException {
+    framework.addInstance(Table.class.getName(), DATASET_ID, DATASET_PROPERTIES);
+  }
+
+  /**
+   * Store the owner's principal for the given {@link EntityId}
+   *
+   * @param entityId The {@link EntityId} whose owner principal needs to be stored
+   * @param principal the principal of the {@link EntityId} owner
+   * @throws IOException if failed to get the store
+   */
+  public void add(final EntityId entityId, final String principal) throws IOException, AlreadyExistsException {
+    if (exists(entityId)) {
+      throw new AlreadyExistsException(String.format("Owner information already exists for entity '%s'.",
+                                                     entityId));
+    }
+    // Before storing it make sure its a valid kerberos principal to fail early and hence failing the entity creation
+    SecurityUtil.validatePrincipal(principal);
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          Table metaTable = getTable(context);
+          metaTable.put(createRowKey(entityId), COL, Bytes.toBytes(principal));
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e, IOException.class);
```

Comment:
When you move the exists check in the same transaction, remember to propagate the `AlreadyExistsException` as well.